src/submodule.c

```diff
@@ -715,6 +720,22 @@ const char *git_submodule_branch(git_submodule *submodule)
 	return submodule->branch;
 }
 
+int git_submodule_set_branch(git_submodule *submodule, const char *branch)
+{
+	assert(submodule);
+
+	git__free(submodule->branch);
+
+	if (branch != NULL) {
+		submodule->branch = git__strdup(branch);
+		GITERR_CHECK_ALLOC(submodule->branch);
+	} else {
+		submodule->branch = NULL;
+	}
```

Comment:
If the strdup fails, we will leave `submodule->branch` pointing to unallocated memory. We should set it to `NULL` always after freeing.